File: src/main_grabcut.cpp
/**
 * @brief  Script to run a GMM based on a trimap or a rect.
 *
 * @date   20 August 2019.
 * @author Luis Carlos Garcia Peraza Herrera (luiscarlos.gph@gmail.com).
 */

#define STRCASECMP strcasecmp
#define STRNCASECMP strncasecmp

#include <chrono>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <iostream>

// CUDA
#include <cuda_runtime.h>
#include <npp.h>

// OpenCV
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>

// My includes
#include <cmdlinereader.h>
#include <grabcut.h>

int main(int argc, char **argv) {
  // Parse command line
  if (!CommandLineReader::getInstance().processCmdLineOptions(argc, argv)) return EXIT_FAILURE;

  // Load image
  const std::string defaultImage = CommandLineReader::getInstance().getInputFilePath();
  cv::Mat image = cv::imread(defaultImage);
  if (!image.data) {
    std::cout << "Could not open or find the image" << std::endl;
    return EXIT_FAILURE;
  }

  // Convert image to BGRA
  cv::Mat bgra;
  cv::cvtColor(image, bgra, cv::COLOR_BGR2BGRA);
#ifndef NDEBUG
  cv::imshow("Original image", bgra);
  cv::waitKey();
#endif

  // Get maximum number of iterations from command line
  int maxIter = CommandLineReader::getInstance().getNumberOfIterations();

  // Load rect image and find rect coordinates
  cv::Mat result;
  cv::Mat probaResult;
  if (!CommandLineReader::getInstance().getRectFilePath().empty()) {
    // The user inputs an image with black background and a white rectangle on
    // the ROI

    cv::Mat rectColor = cv::imread(CommandLineReader::getInstance().getRectFilePath());
    cv::Mat rectGray;
    cv::cvtColor(rectColor, rectGray, cv::COLOR_BGR2GRAY);
#ifndef NDEBUG
    cv::imshow("Rect image loaded from file", rectGray);
    cv::waitKey();
#endif

    // Find rectangle coordinates and convert them to OpenCV rect
    cv::Rect rect;
    rect.y = -1;
    for (int i = 0; i < rectGray.rows; ++i) {
      uchar *p = rectGray.ptr<uchar>(i);
      for (int j = 0; j < rectGray.cols; ++j) {
        if (p[j] > 0 && rect.y == -1) {
          rect.y = i;
          rect.x = j;
        }
        if (p[j] > 0) {
          rect.height = i + 1 - rect.y;
          rect.width = j + 1 - rect.x;
        }
      }
    }

    // Grabcut segmentation
    GrabCut grabcut(maxIter);
    result = grabcut.estimateSegmentationFromRect(bgra, rect);
  } else if (!CommandLineReader::getInstance().getTrimapFilePath().empty()) {
    // We expect the input image in this format:
    //
    //    0   = sure background
    //    128 = unknown
    //    255 = sure foreground
    //
    cv::Mat trimap = 
      cv::imread(CommandLineReader::getInstance().getTrimapFilePath(), cv::IMREAD_GRAYSCALE);

    // Grabcut segmentation
    GrabCut grabcut(maxIter);
    result = grabcut.estimateSegmentationFromTrimap(bgra, trimap);
  } else if (!CommandLineReader::getInstance().getFourmapFilePath().empty()) {
    // We expect the input image in this format:
    //
    //    0   = sure background
    //    64  = probably background
    //    128 = probably foreground
    //    255 = sure foreground
    //
    cv::Mat fourmap =
      cv::imread(CommandLineReader::getInstance().getFourmapFilePath(), cv::IMREAD_GRAYSCALE);

    // Grabcut segmentation
    GrabCut grabcut(maxIter);
    result = grabcut.estimateSegmentationFromFourmap(bgra, fourmap); 
  } else if (!CommandLineReader::getInstance().getFgMapFilePath().empty()) {
    // Load background probability map
    cv::Mat bgMap =
      cv::imread(CommandLineReader::getInstance().getBgMapFilePath(), cv::IMREAD_GRAYSCALE);
    // Load foreground probability map
    cv::Mat fgMap =
      cv::imread(CommandLineReader::getInstance().getFgMapFilePath(), cv::IMREAD_GRAYSCALE);

    // Convert probability maps to [0, 1] range 
    cv::Mat probaBg, probaFg;
    bgMap.convertTo(probaBg, CV_32FC1);
    fgMap.convertTo(probaFg, CV_32FC1);
    probaBg /= 255.0f;
    probaFg /= 255.0f;

    // Perform cut using the probability maps to obtain the unary terms
    GrabCut grabcut(maxIter);
    double gamma = CommandLineReader::getInstance().getGamma();
    if (gamma != -1) {
      result = grabcut.estimateSegmentationFromProba(bgra, probaBg, probaFg, gamma);
    } 
    else {
      result = grabcut.estimateSegmentationFromProba(bgra, probaBg, probaFg);
    }
  } else {
    throw std::invalid_argument("Must provide either a rect, trimap or fourmap.");
  }

  // Save result in output file
  result *= 255;
  cv::imwrite(CommandLineReader::getInstance().getOutputFilePath(), result);
  
  return EXIT_SUCCESS;
}

File: src/python/main_grabcut.py
#!/usr/lib/python
#
# @brief This is a testing example to use the grabcut Python module. There is a comparison with
#        the OpenCV vesion.
#
# @author Luis C. Garcia Peraza Herrera (luiscarlos.gph@gmail.com).
# @date   3 Sep 2019.

import numpy as np
import cv2
import time
import argparse

# My imports
import grabcut


def parse_cmdline_params(parser):

    # Mandatory parameters
    parser.add_argument(
        '--image',
        required=True,
        help='Path to the input RGB image.'
    )

    parser.add_argument(
        '--output',
        required=True,
        help='Path to the output segmentation map.'
    )

    parser.add_argument(
        '--iter',
        required=True,
        help='Maximum number of GrabCut iteraions.'
    )

    parser.add_argument(
        '--gamma',
        required=True,
        help='GrabCut gamma parameter. Typical value 50.0.'
    )

    # Optional parameters
    parser.add_argument(
        '--trimap',
        required=False,
        default=None,
        help='Path to the input trimap.'
    )

    parser.add_argument(
        '--fourmap',
        required=False,
        default=None,
        help='Path to the input fourmap.'
    )

    parser.add_argument(
        '--fgmap',
        required=False,
        default=None,
        help='Path to the foreground probability map. [0, 255] will be mapped to [0, 1].'
    )

    parser.add_argument(
        '--bgmap',
        required=False,
        default=None,
        help='Path to the background probability map. [0, 255] will be mapped to [0, 1].'
    )

    return parser.parse_args()

#
# @brief Validates the input parameters given by the user in the command line.
#


def validate_cmdline_params(args):
    assert(args.trimap or args.fourmap or args.fgmap)

#
# @brief Converts the command line arguments to the right data types.
#


def convert_cmdline_params(args):
    args.iter = int(args.iter)
    args.gamma = float(args.gamma)


def main():
    # Parse command line parameters
    parser = argparse.ArgumentParser()
    args = parse_cmdline_params(parser)
    validate_cmdline_params(args)
    convert_cmdline_params(args)

    # Read image
    im = cv2.imread(args.image)
    im = cv2.cvtColor(im, cv2.COLOR_BGR2BGRA)

    # Create GrabCut algo
    gc = grabcut.GrabCut(args.iter)

    # Compute segmentation
    seg = None
    if args.trimap is not None:
        # Read trimap
        trimap = cv2.imread(args.trimap)[:, :, 0]

        # Compute segmentation
        tic = time.time()
        seg = gc.estimateSegmentationFromTrimap(im, trimap, args.gamma)
        toc = time.time()

    elif args.fourmap is not None:
        # Read fourmap
        fourmap = cv2.imread(args.fourmap)[:, :, 0]

        # Compute segmentation
        tic = time.time()
        seg = gc.estimateSegmentationFromFourmap(im, fourmap, args.gamma)
        toc = time.time()

    elif args.fgmap is not None:
        # Read foreground map
        fgmap = cv2.imread(args.fgmap)[:, :, 0].astype(np.float32) / 255.0
        
        # Read or compute background map
        bgmap = None
        if args.bgmap is not None:
            bgmap = cv2.imread(args.bgmap)[:, :, 0].astype(np.float32) / 255.0
        else:
            bgmap = 1.0 - fgmap;

        # Compute segmentation 
        tic = time.time()
        seg = gc.estimateSegmentationFromProba(im, bgmap, fgmap, args.gamma)
        toc = time.time()

    # Save output and print timing
    seg *= 255  # the output from GrabCut is (0, 1)
    cv2.imwrite(args.output, seg)
    print('Time elapsed in GrabCut segmentation: ' + str(toc - tic))

if __name__ == "__main__":
    main()
